adds support for id_column in tables for auto_publish

Binabh opened the pull request:
Resolves #682 

- [x] Get id_column string from config.yaml and use for id column
- [x] Support for list of strings
- [x] Add info/warnings if column is not there or is of wrong type
- [x] if column for the feature ID is found, remove it from properties (see inline comment)
- [x] cleanup logging messages
- [x] need more tests to catch other edge cases

Binabh committed:
adds support for id_column in tables

Binabh committed:
fix rust suggestions

Binabh committed:
remove unnecessary call to map

Binabh committed:
add support for multiple id columns

Binabh committed:
adds info for id column not found

Binabh committed:
convert match to if let

nyurik commented on martin/src/pg/table_source.rs:
```diff
@@ -193,6 +203,38 @@ FROM {schema}.{table};
         .and_then(|p| polygon_to_bbox(&p)))
 }
 
+async fn calc_id_column(
```
I don't think you should query DB for the available IDs. `TableInfo.properties` already contains all available columns and their types - you can simply look through that HashMap and see if any of the requested ID columns are present.

nyurik commented on martin/src/pg/configurator.rs:
```diff
@@ -265,6 +276,21 @@ fn new_auto_publish(config: &PgConfig, is_function: bool) -> Option<PgBuilderPub
+        },
+        None => None,
+    }
+}
```
I think this would be more readable (technically it can be one big destructuring statement instead of two, but this is more readable I think. I am not certain if this func should return an optional vector of strings or not -- ` OneOrMany` has a number of methods to convert it to a slice, etc.

```rust
fn get_column_id(config: &PgConfig) -> Option<Vec<String>> {
    if let Some(BoolOrObject::Object(v)) = &config.auto_publish {
        if let Some(BoolOrObject::Object(v)) = &v.tables {
            return v.id_column.as_ref().map(|v| v.iter().cloned().collect());
        }
    }
    None
}
```

nyurik reviewed (commented):
@Binabh thank you so much for working on this!  Left a few suggestions inline

Binabh committed:
change logic to check available columns and types

Binabh committed:
collapse if else block

Binabh committed:
rename id_formatter to source_id_formatter with backwards compatibility

nyurik committed:
minor cleanup suggestions

* don't clone when you can avoid it
* i think vec<T> is a bit easier here - just an internal optimization,
  shouldn't be exposed like this (not a clean external API, but ok
internally)

Binabh committed:
Merge pull request #1 from nyurik/column-id-fixes

nyurik committed:
Merge branch 'main' into specify-column-id-auto-publish

nyurik committed:
minor changes to the id/typ check

nyurik committed:
Merge branch 'main' into specify-column-id-auto-publish

nyurik committed:
Merge branch 'main' into specify-column-id-auto-publish

nyurik committed:
Merge branch 'main' into specify-column-id-auto-publish

nyurik committed:
Partial implementation of the id_column detection

Detecting id column should only be done when adding all auto-detected tables to the sources.

I rewrote the logic, but did not test it fully. I added a simple test (we may want to add a few more), but currently tests break - not certain why yet - maybe the `from_schema` is not working correctly?

It's getting a bit late - try going from here.

Binabh committed:
Merge pull request #2 from nyurik/specify-column-id-auto-publish

nyurik committed:
Merge branch 'main' into specify-column-id-auto-publish

Also fixed tests and other minor cleanup

nyurik commented on tests/expected/given_config.yaml:
```diff
@@ -28,6 +29,7 @@ postgres:
       table: auto_table
       srid: 4326
       geometry_column: geom
+      id_column: feat_id
```
the `feat_id: int4` line in the `properties` section on line 43 should not be there -- it indicates that the same value from the `feat_id` column will be included twice - once as the feature ID, and again as a regular propery.

nyurik commented on docs/src/config-file.md:
```diff
@@ -67,6 +67,10 @@ postgres:
+      # A table column to use as the feature ID
+      # If a table has no column with this name, `id_column` will not be set for that table.
+      # If a list of strings is given, the first found column will be treated as a feature ID.
+      id_column: feature_id
```
This should probably be `id_columns`?  Note that both are allowed because of the Serde's `alias` attribute.

nyurik commented on martin/src/pg/configurator.rs:
```diff
@@ -232,12 +234,54 @@ impl PgBuilder {
+        } else {
+            match find_kv_ignore_case(props, key) {
+                Ok(Some(result)) => {
+                    info!("For source {id} from table {}.{}, feature ID '{key}' was not found, but found '{result}' instead.", inf.schema, inf.table);
```
the logging here and below might need to be cleaned up to make it consistent with the `pg::utilities` logging, and include the source ID in all, etc.

nyurik commented on martin/src/pg/configurator.rs:
```diff
@@ -232,12 +234,54 @@ impl PgBuilder {
+        };
+        // ID column can be any integer type as defined in
+        // https://github.com/postgis/postgis/blob/559c95d85564fb74fa9e3b7eafb74851810610da/postgis/mvt.c#L387C4-L387C66
+        if typ != "int4" && typ != "int8" && typ != "int2" {
```
need to double check if PG can ever return `int` or some other types here?  Likely not, just a thought.

Binabh replied:
According to : https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-INT We have covered for all integer types.

nyurik commented on tests/config.yaml:
```diff
@@ -22,6 +22,7 @@ postgres:
   auto_publish:
     tables:
       from_schemas: autodetect
+      id_column: feat_id
```
```suggestion
      id_columns: feat_id
```

nyurik reviewed (commented):
a few todos inline

Binabh committed:
rename id_column to id_columns

Binabh committed:
make logging consistent

Binabh committed:
remove duplicate id_column value in properties

Binabh committed:
remove column from properties if already in id_column

Binabh commented on martin/src/pg/configurator.rs:
```diff
@@ -232,12 +234,57 @@ impl PgBuilder {
+        }
+
+        inf.id_column = Some(column.to_string());
+        let mut final_props = props.clone();
```
@nyurik as you said earlier that we should avoid clone whenever possible. can this clone be avoided?

nyurik replied:
it is more of a "don't clone unless you have to" :)

I agree that here it would be better here without the clone, but the issue is that you have a mutable ref to a hashmap - `props`, plus you have a readonly ref to one of the keys in that hashmap - `column` - and Rust won't let you have both readonly and mutable refs to the same object at the same time.  If you can figure out a way around that, sure :)

Binabh replied:
Yeah, I also searched a little and found that this was only way to do it. So, for now let's resolve it as it is.

Binabh committed:
add test for bigint feature id

nyurik committed:
minor test cleanup

nyurik committed:
Merge branch 'main' into specify-column-id-auto-publish

nyurik reviewed (approved):
thanks!!!
